### DISCUSSION TITLE
Using Tree-sitter CSTs in a VSCode Extension

### DISCUSSION THREAD

**Discussion by tinygiant98** [tree-sitter/tree-sitter#1162] [Q&A] [answered]
As I attempt to write my first extension for VSCode, I'm trying to incorporate tree-sitter to parse and "new" language.  The integration with VSCode has good semi-swimmingly so far.  I've been able to integrate the new language and the appropriate events fire (activate, etc.) when the document opens and is edited.  Token coloring occurs where it should and the colors match what I've got setup for token mapping.

However, VSCode refuses to display errors that are present in the CST.  I can view the CST of a specific file and I know there are either ERROR or MISSING nodes, but the standard red/orange squiggle underline is missing.  I can't seem to find any reference to options for enabling it and the underline seems to work for all other semantic token providers in other extensions.  I have to assume it's automatic when a specific node is encountered in the syntax tree.  There are also no issues listed in the "problems" 'section of the terminal, so it seems as though vscode doesn't not recognize the problem nodes.  Anyone have any experience with this?

**Comment by ahlinc** [comment, marked as answer]
VSCode doesn't do anything by its own and doesn't have any builtin facilities to somehow _"automatically"_ understand any AST/CST, it's an extension author responsibility to construct some _"objects/entities"_ through VSCode API to represent meaningful things.
For your case you need to construct the `DiagnosticCollection` by https://code.visualstudio.com/api/references/vscode-api#languages.createDiagnosticCollection and manage all related things for by yourself.

**Reply by tinygiant98** [reply]
Thanks! I discovered diagnostics about an hour ago and started hacking on it.  Hopefully I can get a working example soon, just need to figure out the most efficient place to hook into the token update process. 

Thanks again!

**Reply by tinygiant98** [reply]
Is there any documentation on fields that are available from each CST node as it's traversed? If not, which module should I start hunting in to find the available fields?

**Reply by ahlinc** [reply]
I suppose for your case it's `d.ts` file for WASM binding:
https://github.com/tree-sitter/tree-sitter/blob/6ed42747a4e0faee9b65edbbacc86ed0caeae05c/lib/binding_web/tree-sitter-web.d.ts#L50-L95
And match it with documentation: https://tree-sitter.github.io/tree-sitter/using-parsers#syntax-nodes which is primarily for C API.

**Reply by razzeee** [reply]
Be aware of https://github.com/tree-sitter/tree-sitter/issues/493

**Reply by tinygiant98** [reply]
> Be aware of https://github.com/tree-sitter/tree-sitter/issues/493

Thanks for the warning!

**Reply by tinygiant98** [reply]
@ahlinc ... Wow, wasn't expecting that much data to sift through/work with.   Should be able to create some interesting features with access to all that. Thanks!

**Reply by tinygiant98** [reply]
Is there any ability to pass custom data to a node during the parsing process (besides the type that it parses as)?

**Reply by ahlinc** [reply]
> Is there any ability to pass custom data to a node during the parsing process (besides the type that it parses as)?

No, it isn't possible, there are [many](https://github.com/tree-sitter/tree-sitter/issues/1085#issuecomment-830654467) [reasons](https://github.com/tree-sitter/tree-sitter/discussions/1108#discussioncomment-842884) for this.
